### PR TITLE
Implement O(1)-time Iterator::nth for Range*, and slim the Step trait

### LINKED ISSUE #43064: The new step_by is too much slow
As I explained in #42534 the new step_by can't be used:

```
#![feature(iterator_step_by)]

#[inline(never)]
pub fn first_foo(mut a: i32, b: i32) -> i32 {
    let mut tot = 0;

    while a < b {
        tot += a;
        a += 3;
    }
    tot
}

#[inline(never)]
pub fn second_foo(x: std::ops::Range<i32>) -> i32 {
    let mut tot = 0;

    #[allow(deprecated)]
    for a in x.step_by(3) {
        tot += a;
    }
    tot
}

fn main() {
    println!("{}", first_foo(0, 1000));
    println!("{}", first_foo(0, 100000));
    println!("{}", second_foo(0 .. 1000));
    println!("{}", second_foo(0 .. 100000));
}
```

Compiling with:
rustc -O --emit asm test.rs

It gives:

```
_ZN4test9first_foo17he3a0bddb42457919E:
    testl   %ecx, %ecx
    jle .LBB0_1
    xorl    %edx, %edx
    xorl    %eax, %eax
    .p2align    4, 0x90
.LBB0_4:
    addl    %edx, %eax
    addl    $3, %edx
    cmpl    %ecx, %edx
    jl  .LBB0_4
    jmp .LBB0_2
.LBB0_1:
    xorl    %eax, %eax
.LBB0_2:
    retq

_ZN4test10second_foo17hd5d21559ba84e6d1E:
    movq    %rcx, %r9
    shrq    $32, %r9
    xorl    %r8d, %r8d
    xorl    %r10d, %r10d
    xorl    %eax, %eax
    testb   $1, %r10b
    je  .LBB1_11
    jmp .LBB1_2
    .p2align    4, 0x90
.LBB1_7:
    shrq    $32, %rdx
    addl    %eax, %edx
    movb    $1, %r10b
    movl    %edx, %eax
    testb   $1, %r10b
    je  .LBB1_11
.LBB1_2:
    cmpl    %r9d, %ecx
    jge .LBB1_5
    leal    1(%rcx), %edx
    cmpl    %r9d, %edx
    jge .LBB1_4
    leal    2(%rcx), %edx
    cmpl    %r9d, %edx
    jge .LBB1_9
    addl    $3, %ecx
    shlq    $32, %rdx
    movl    $1, %r10d
    jmp .LBB1_6
    .p2align    4, 0x90
.LBB1_11:
    movq    %rcx, %r10
    shlq    $32, %r10
    xorl    %edx, %edx
    cmpl    %r9d, %ecx
    setl    %dl
    cmovgeq %r8, %r10
    addl    %edx, %ecx
    jmp .LBB1_6
    .p2align    4, 0x90
.LBB1_4:
    movl    %edx, %ecx
    jmp .LBB1_5
.LBB1_9:
    movl    %edx, %ecx
    .p2align    4, 0x90
.LBB1_5:
    xorl    %edx, %edx
    xorl    %r10d, %r10d
.LBB1_6:
    orq %r10, %rdx
    testl   %edx, %edx
    jne .LBB1_7
    retq
```

In a function like second_foo() the old step_by gave something like:

```
_ZN5test210second_foo17h5b0eb3168418bdd1E:
    movq    %rcx, %rdx
    shrq    $32, %rdx
    xorl    %eax, %eax
    cmpl    %edx, %ecx
    jge .LBB1_3
    xorl    %eax, %eax
    .p2align    4, 0x90
.LBB1_2:
    addl    %ecx, %eax
    addl    $3, %ecx
    cmovol  %edx, %ecx ; what is this for?
    cmpl    %edx, %ecx
    jl  .LBB1_2
.LBB1_3:
    retq
```

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes #43064.
Fixes part of #39975.
Fixes items 1 <s>and 3</s> of #42168.
CC #27741.

I think #42310 and #43012 should not have landed without the `nth` part of this PR, but oh well.

### CODE REVIEW

File: src/libcore/iter/range.rs [hunk 1]
```diff
@@ -64,6 +67,16 @@ macro_rules! step_identical_methods {
         fn sub_one(&self) -> Self {
             Sub::sub(*self, 1)
         }
+
+        #[inline]
+        fn add_n(&self, n: usize) -> Option<Self> {
+            // FIXME: use faillable conversions when they exist
```
Comment (nox) [inline src/libcore/iter/range.rs]: Fallible.

**Comment by kennytm** [comment]
A test case needs to be updated.

```rust
[00:40:46] ---- [compile-fail] compile-fail/range-1.rs stdout ----
[00:40:46] 	
[00:40:46] error: /checkout/src/test/compile-fail/range-1.rs:19: expected error not found: `for<'a> &'a bool: std::ops::Add` is not satisfied
```

File: src/libcore/iter/range.rs (hunk 1 above)
Comment (SimonSapin) [inline src/libcore/iter/range.rs]: Fixed.

**Comment by SimonSapin** [comment]
Test updated, thanks for pointing this out.

**Comment by kennytm** [comment]
Need to update [`run-pass/impl-trait/example-calendar.rs`](https://github.com/rust-lang/rust/blob/master/src/test/run-pass/impl-trait/example-calendar.rs#L164) too :joy:

<details><summary>Error report</summary>

```
[00:52:28] ---- [run-pass] run-pass/impl-trait/example-calendar.rs stdout ----
[00:52:28] 	
[00:52:28] error: compilation failed!
[00:52:28] status: exit code: 101
[00:52:28] command: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc /checkout/src/test/run-pass/impl-trait/example-calendar.rs -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass --target=x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/impl-trait/example-calendar.stage2-x86_64-unknown-linux-gnu.run-pass.libaux -C prefer-dynamic -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/impl-trait/example-calendar.stage2-x86_64-unknown-linux-gnu -Crpath -O -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers
[00:52:28] stdout:
[00:52:28] ------------------------------------------
[00:52:28] 
[00:52:28] ------------------------------------------
[00:52:28] stderr:
[00:52:28] ------------------------------------------
[00:52:28] error[E0407]: method `step` is not a member of trait `std::iter::Step`
[00:52:28]    --> /checkout/src/test/run-pass/impl-trait/example-calendar.rs:165:5
[00:52:28]     |
[00:52:28] 165 | /     fn step(&self, by: &Self) -> Option<Self> {
[00:52:28] 166 | |         Some(self + by)
[00:52:28] 167 | |     }
[00:52:28]     | |_____^ not a member of trait `std::iter::Step`
[00:52:28] 
[00:52:28] error[E0407]: method `steps_between_by_one` is not a member of trait `std::iter::Step`
[00:52:28]    --> /checkout/src/test/run-pass/impl-trait/example-calendar.rs:173:5
[00:52:28]     |
[00:52:28] 173 | /     fn steps_between_by_one(_: &Self, _: &Self) -> Option<usize> {
[00:52:28] 174 | |         unimplemented!()
[00:52:28] 175 | |     }
[00:52:28]     | |_____^ not a member of trait `std::iter::Step`
[00:52:28] 
[00:52:28] error[E0407]: method `is_negative` is not a member of trait `std::iter::Step`
[00:52:28]    --> /checkout/src/test/run-pass/impl-trait/example-calendar.rs:177:5
[00:52:28]     |
[00:52:28] 177 | /     fn is_negative(&self) -> bool {
[00:52:28] 178 | |         false
[00:52:28] 179 | |     }
[00:52:28]     | |_____^ not a member of trait `std::iter::Step`
[00:52:28] 
[00:52:28] error[E0050]: method `steps_between` has 3 parameters but the declaration in trait `std::iter::Step::steps_between` has 2
[00:52:28]    --> /checkout/src/test/run-pass/impl-trait/example-calendar.rs:169:45
[00:52:28]     |
[00:52:28] 169 |     fn steps_between(_: &Self, _: &Self, _: &Self) -> Option<usize> {
[00:52:28]     |                                             ^^^^^ expected 2 parameters, found 3
[00:52:28]     |
[00:52:28]     = note: `steps_between` from trait: `fn(&Self, &Self) -> std::option::Option<usize>`
[00:52:28] 
[00:52:28] error[E0046]: not all trait items implemented, missing: `add_n`
[00:52:28]    --> /checkout/src/test/run-pass/impl-trait/example-calendar.rs:164:1
[00:52:28]     |
[00:52:28] 164 | / impl std::iter::Step for NaiveDate {
[00:52:28] 165 | |     fn step(&self, by: &Self) -> Option<Self> {
[00:52:28] 166 | |         Some(self + by)
[00:52:28] 167 | |     }
[00:52:28] ...   |
[00:52:28] 195 | |     }
[00:52:28] 196 | | }
[00:52:28]     | |_^ missing `add_n` in implementation
[00:52:28]     |
[00:52:28]     = note: `add_n` from trait: `fn(&Self, usize) -> std::option::Option<Self>`
[00:52:28] 
[00:52:28] error: aborting due to 5 previous errors
[00:52:28] 
[00:52:28] 
[00:52:28] ------------------------------------------
[00:52:28] 
[00:52:28] thread '[run-pass] run-pass/impl-trait/example-calendar.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2473:8
[00:52:28] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:52:28] 
[00:52:28] 
[00:52:28] failures:
[00:52:28]     [run-pass] run-pass/impl-trait/example-calendar.rs
```

</details>

**Comment by SimonSapin** [comment]
Removed the i128/u128 changes as I suspect they were buggy. I’ll look into that separately in another PR.

**Comment by alexcrichton** [comment]
Thanks! Could you be sure to add some tests for the `nth` overrides here as well?

**Comment by SimonSapin** [comment]
Done, in a new commit.

**Comment by alexcrichton** [comment]
@bors: r+

cc @scottmcm

**Pull request merged**